### PR TITLE
refactor: introduce a newtype for vnode

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

To avoid leaking the logic to the upper layer, and makes specifying a new vnode count easier.

https://github.com/risingwavelabs/risingwave/blob/ce22f0c2b61d082ab493b63df85262c64c561e04/src/common/src/hash/vnode.rs#L24-L28

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #6016

### CODE REVIEW

**Review by yezizp2012 — APPROVED** [review]
LGTM

File: src/common/src/hash/vnode.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    /// Returns the big-endian bytes representation of the virtual node.
+    pub const fn to_be_bytes(self) -> [u8; Self::SIZE] {
+        self.0.to_be_bytes()
+    }
+
+    /// Returns the next virtual node.
+    pub const fn next(self) -> Self {
+        Self(self.0 + 1)
+    }
```
Comment (xx01cyx) [inline src/common/src/hash/vnode.rs:90]: Should we check for overflow here?

**Review by TennyZhuang — APPROVED** [review]

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 2]
```diff
@@ -283,19 +289,14 @@ impl<'a> UnorderedRangeCacheIter<'a> {
     }
 
     fn refill_iterator(&mut self) {
-        loop {
-            if self.vnodes.is_set(self.next_vnode as usize) && let Some(vnode_range) = self.cache.get(&self.next_vnode) {
+        while let Some(vnode) = self.vnode_iter.next() {
+            if self.vnodes.is_set(vnode.to_index()) && let Some(vnode_range) = self.cache.get(&vnode) {
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:293]: Am I correct here? PTAL @jon-chuang

File: src/common/src/hash/vnode.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/hash/vnode.rs]: I've added debug assertions for all initializations and removed this function, as the only usage in the dynamic filter can be covered by `all() -> impl Iterator`.

**Review by xx01cyx — APPROVED** [review]
LGTM

**Review by jon-chuang — APPROVED** [review]
LGTM

**Pull request merged**